service/notification/internal/broker/nats_broker.go

```diff
@@ -0,0 +1,129 @@
+package broker
+
+import (
+	"fmt"
+
+	"context"
+	"time"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-boredgus/service/notification/internal/config"
+	"github.com/google/uuid"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+type (
+	ConsumedMessage interface {
+		// Subject returns a subject on which a message was published/received.
+		Subject() string
+
+		// Data returns the message body.
+		Data() []byte
+
+		// Ack tells the server that the message was successfully processed
+		// and it can move on to the next message.
+		Ack() error
+
+		// NakWithDelay tells the server to redeliver the message after the given delay.
+		NakWithDelay(delay time.Duration) error
+	}
+
+	natsBroker struct {
+		jetstream jetstream.JetStream
+		consumer  jetstream.Consumer
+		logger    config.Logger
+	}
+)
+
+const (
+	CreationTimeout time.Duration = 5 * time.Second
+
+	NotificationStreamName   string = "NOTIFICATIONS"
+	NotificationConsumerName string = "notification-sender"
+)
+
+func createConsumer(js jetstream.JetStream) (jetstream.Consumer, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), CreationTimeout)
+	defer cancel()
+
+	stream, err := js.CreateOrUpdateStream(ctx, jetstream.StreamConfig{
+		Name:      NotificationStreamName,
+		Retention: jetstream.WorkQueuePolicy,
+		Subjects:  []string{"commands.>", "events.subscription.>"},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update %s stream: %w", NotificationStreamName, err)
+	}
+
+	consumer, err := stream.CreateOrUpdateConsumer(context.Background(), jetstream.ConsumerConfig{
+		Name:          NotificationConsumerName,
+		Durable:       NotificationConsumerName,
+		DeliverPolicy: jetstream.DeliverAllPolicy,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update %s consumer: %w", NotificationConsumerName, err)
+	}
+
+	return consumer, nil
+}
+
+func NewNatsBroker(js jetstream.JetStream, logger config.Logger) (*natsBroker, error) {
+	consumer, err := createConsumer(js)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init consumer: %w", err)
+	}
```

Comment:
I guess this shouldn't be at consructor, but relied more on client to register subscribers

Reply:
Moved from constructor to entrypoint of service